### PR TITLE
Pulsar connector

### PR DESCRIPTION
**Pull request by ozangunalp**
Thanks to the work from @geoand 

Still more tests to be done for threading and different client configuration scenarios.

### CODE REVIEW

**Comment by geoand** [comment]
🎉

**Comment by cescoffier** [comment]
Unfortunately, we want to avoid having reactor on the classpath. It will bring confusion, as Quarkus, the main consumer, only integrates mutiny.

**Comment by scrocquesel** [comment]
Hi I know this is still work in progress, but I'm trying to build a quarkus extension on top of this great work to provide Dev services XP. I'm currently stuck at testing because no message arrives at the broker. Both consumer and producer are connected to the broker. I'm using the exact same test classes used in integration tests of mqtt or rabbitmq quarkus extensions.

```java
    @Inject
    @Channel("people-out")
    Emitter<String> emitter;

    public void seedPeople() {
        Stream
                .of("bob",
                        "alice",
                        "tom",
                        "jerry",
                        "anna",
                        "ken")
                .forEach(s -> emitter.send(s));
    }
```

```properties
mp.messaging.outgoing.people-out.connector=smallrye-pulsar
mp.messaging.outgoing.people-out.topic=people
```

When calling the `seedPeople` method, nothing seems to happen.

repro at https://github.com/scrocquesel/smallrye-reactive-messaging-pulsar

Should you have a look ?

**Comment by ozangunalp** [comment]
Hi, I've actually started working on the Quarkus extension.
I needed to iron out some issues injecting config but other than that it seems fine on JVM.
( For your example, I think you'll need to set schema on your channel like `mp.messaging.outgoing.people-out.schema=STRING`)
I am working to enable some kind of schema auto-generation.

I expect more resistance for native-mode.

**Comment by scrocquesel** [comment]
> ( For your example, I think you'll need to set schema on your channel like `mp.messaging.outgoing.people-out.schema=STRING`) 

Thanks, I'm discovering Pulsar and didn't know it is that schema oriented.

> I am working to enable some kind of schema auto-generation.

That would be great. I also test receiving `byte[]` and I had to use `GenericRecord` and casting the native object to `byte[]`. I guess by default you use auto schema, and tha's why the receiver got a `GenericRecord`. I don't know if it is possible nor desirable but could smallrye wrap this complexity for `byte[]` and `string` when no schema is specified ?

> I expect more resistance for native-mode.

For now, I'm focusing more on dev service experience. I'm discovering Pulsar and I miss the management UI from the stock Pulsar image. I would like to start the management UI aside or configure automatically a running one. Let me know if you are interested.

File: documentation/src/main/docs/pulsar/receiving-pulsar-messages.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+```properties
+mp.messaging.incoming.prices.connector=smallrye-pulsar # <1>
+mp.messaging.incoming.prices.serviceUrl=pulsar://pulsar:6650 # <2>
+mp.messaging.incoming.prices.schema=DOUBLE # <3>
+mp.messaging.incoming.prices.subscriptionInitialPosition=Earliest # <4>
+```
+
+1.  Sets the connector for the `prices` channel
+
```
Comment (cescoffier) [inline documentation/src/main/docs/pulsar/receiving-pulsar-messages.md:21]: you sure about the extra blank line here?

File: documentation/src/main/docs/pulsar/receiving-pulsar-messages.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+It provides the key, topic, partitions, headers and so on:
+
+```java
+{{ insert('pulsar/inbound/PulsarMetadataExample.java', 'code') }}
+```
+
+## Acknowledgement
+
+When a message produced from a Pulsar Message is *acknowledged*, the connector sends an [acknowledgement request](https://pulsar.apache.org/docs/3.0.x/concepts-messaging/#acknowledgment) to the Pulsar broker.
+All Reactive Messaging messages needs to be *acknowledged*, and in most cases this is handled automatically.
```
Comment (cescoffier) [inline documentation/src/main/docs/pulsar/receiving-pulsar-messages.md:75]:
```suggestion
All Reactive Messaging messages need to be *acknowledged*, which is handled automatically in most cases.
```

File: documentation/src/main/docs/pulsar/sending-messages-to-pulsar.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+the `prices` channel into a Pulsar Messages as follows:
+
+```properties
+mp.messaging.outgoing.prices.connector=smallrye-pulsar # <1>
+mp.messaging.outgoing.prices.serviceUrl=pulsar://pulsar:6650 # <2>
+mp.messaging.outgoing.prices.schema=DOUBLE # <3>
+```
+
+1.  Sets the connector for the `prices` channel
+
```
Comment (cescoffier) [inline documentation/src/main/docs/pulsar/sending-messages-to-pulsar.md:19]: are you sure about the extra blank line?

File: documentation/src/main/docs/pulsar/client-configuration.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+You also can provide a `Map<String, Object>` containing configuration values by key:
+
+```java
+{{ insert('pulsar/configuration/PulsarClientConfigProducers.java', 'producer') }}
+```
+
+
+In the order of precedence, the least to the highest, each step overriding the previous one,
+following are the different sources of configuration:
```
[the remark this answers is not in the document]
Comment (ozangunalp) [reply documentation/src/main/docs/pulsar/client-configuration.md]: Replaced with `Different configuration sources are loaded in the following order of precedence, from the least important to the highest:`

**Review by cescoffier — APPROVED** [review]

**Comment by cescoffier** [comment]
Woohoo! 

I just approved.

@ozangunalp I let you merge, release and open the Quarkus PR.

**Comment by geoand** [comment]
Amazing!!!

**Comment by jponge** [comment]
NIce work!

**Review by cescoffier — APPROVED** [review]

**Pull request merged**